cmd/zoekt-sourcegraph-indexserver/queue.go

```diff
@@ -228,46 +338,130 @@ type pqueue []*queueItem
 func (pq pqueue) Len() int { return len(pq) }
 
 func (pq pqueue) Less(i, j int) bool {
-	// If we know x needs an update and y doesn't, then return true. Otherwise
-	// they are either equal priority or y is more urgent.
 	x := pq[i]
 	y := pq[j]
-	if x.indexed != y.indexed {
-		return !x.indexed
-	}
-
-	if xFail, yFail := x.indexState == indexStateFail, y.indexState == indexStateFail; xFail != yFail {
-		// if you failed to index, you are likely to fail again. So prefer
-		// non-failed.
-		return !xFail
-	}
-
-	// tie breaker is to prefer the item added to the queue first
-	return x.seq < y.seq
+	return lessQueueItemPriority(x, y)
 }
 
 func (pq pqueue) Swap(i, j int) {
 	pq[i], pq[j] = pq[j], pq[i]
-	pq[i].heapIdx = i
-	pq[j].heapIdx = j
+	pq[i].HeapIdx = i
+	pq[j].HeapIdx = j
 }
 
 func (pq *pqueue) Push(x interface{}) {
 	n := len(*pq)
 	item := x.(*queueItem)
-	item.heapIdx = n
+	item.HeapIdx = n
 	*pq = append(*pq, item)
 }
 
 func (pq *pqueue) Pop() interface{} {
 	old := *pq
 	n := len(old)
 	item := old[n-1]
-	item.heapIdx = -1
+	item.HeapIdx = -1
 	*pq = old[0 : n-1]
 	return item
 }
 
+// lessQueueItemPriority returns true if the indexing priority x is less than that of y.
+func lessQueueItemPriority(x, y *queueItem) bool {
+	// If we know x needs an update and y doesn't, then return true. Otherwise
+	// they are either equal priority or y is more urgent.
+	if x.Indexed != y.Indexed {
+		return !x.Indexed
+	}
+
+	if xFail, yFail := x.IndexState == indexStateFail, y.IndexState == indexStateFail; xFail != yFail {
+		// if you failed to index, you are likely to fail again. So prefer
+		// non-failed.
+		return !xFail
+	}
+
+	// tiebreaker is to prefer the item added to the queue first
+	return x.Seq < y.Seq
+}
+
+// queueItemStreamDecoder processes streams of gob-encoded queueItems.
+type queueItemStreamDecoder struct {
```

Comment:
While writing this abstraction, it occurred to me that there is nothing really tying this implementation to only queue items. I think we could make a generic `gobstream` utility package or something similar to provide a `encoder / decoder` abstraction around passing around an unbounded stream of generic items. Thoughts?